### PR TITLE
CSS-4170 Add TLS relation and Ranger plugin

### PR DESCRIPTION
**Pull request by AmberCharitos**
This PR adds the TLS relation to the Trino Charm (`src/tls.py`) for the creation of `keystore.p12` and `truststore.jks` required by Trino to enable HTTPS. It also provides the methods for enabling the ranger plug-in for Trino (checking existence of `tar` file, unpacking and starting the plug-in). Currently the `policy-mgr-url` needs to be input by the user and the `trino-plugin.tar.gz` is already in the Trino image, in future these should be provided by the Ranger charm. 

To be added in next PR: 

- Automatic enabling of HTTPS without need for `https-enabled` config value. (added to this PR)

- Integration & unit tests.

### CODE REVIEW

File: config.yaml [hunk 1]
```diff
@@ -14,18 +14,33 @@ options:
...
     description: |
       Client id from Google Oauth setup
-    default: "not-provided"
     type: string
   google-client-secret:
     description: |
       Client password from Google Oauth setup
-    default: "not-provided"
+    type: string
+  ranger-acl:
```
Comment (kelkawi-a) [inline config.yaml:32]: Perhaps ```ranger-acl-enabled``` would make it clearer?

File: config.yaml
Comment (kelkawi-a) [inline config.yaml:1]: EOF on new line for all files please.

File: src/charm.py [hunk 2]
```diff
@@ -244,20 +252,81 @@ def _on_restart(self, event):
...
+
+        Raises:
+            ValueError: In case no Oauth credentials are provided
+            RuntimeError: In case keystore does not exist
+        """
+        google_id = self.config.get('google-client-id')
+        google_secret = self.config.get('google-client-secret')
+
+        if google_id is None or google_secret is None:
+            raise ValueError("Oauth credentials not provided")
```
Comment (kelkawi-a) [inline src/charm.py:272]: Maybe specify which config values are missing for clarity?

File: src/charm.py [hunk 3]
```diff
@@ -292,28 +361,39 @@ def _update(self, event):
...
+                return
+        else:
+            config_context = {"HTTPS_ENABLED": False}
+
+        self._push_file(container, config_context, CONFIG_JINJA, CONFIG_PATH)
+
+        if self.config['ranger-acl'] is True:
+            try:
+                self._configure_ranger_plugin(container)
+                command = "/trino-entrypoint.sh"
```
Comment (kelkawi-a) [inline src/charm.py:380]: Should this command be returned by ```_configure_ranger_plugin```?

File: src/charm.py [hunk 4]
```diff
@@ -292,28 +361,39 @@ def _update(self, event):
...
             "summary": "trino layer",
             "description": "pebble config layer for trino",
             "services": {
                 self.name: {
                     "override": "replace",
                     "summary": "trino server",
-                    "command": "/usr/lib/trino/bin/run-trino",
+                    "command": command,
                     "startup": "enabled",
-                    "environment": config_env,
```
Comment (kelkawi-a) [inline src/charm.py:316]: Just to confirm, the environment variables are no longer needed when starting the application?

File: src/tls.py
Comment (kelkawi-a) [inline src/tls.py:1]: For future-proofing, these operations should only be done by the leader unit, so you may want to keep in mind that check (Similar to what is done in the [kafka-k8s-operator](https://github.com/canonical/kafka-k8s-operator/blob/1466aecf9fb88ce7f2b7b3a1254377ee760b4ab7/src/tls.py#L55))

File: README.md [hunk 5]
```diff
@@ -1,11 +1,45 @@
 # trino-k8s-charm
-This is the Kubernetes Python Operator for Trino
+The Charmed Trino K8s Operator delivers automated management on Trino data vitulization software on top of a Kubernetes cluter. Trino is a distributed SQL query engine designed to query large data sets distributed over one or more heterogeneous data sources.
```
Comment (kelkawi-a) [inline README.md:2]: data virtualization* and Kubernetes cluster*

File: README.md [hunk 6]
```diff
@@ -1,11 +1,45 @@
...
 
-Trino is a distributed SQL query engine designed to query large data sets distributed over one or more heterogeneous data sources.
+The operator charm comes with features such as:
+- Access control management supported with Ranger provided ACLs
```
Comment (kelkawi-a) [inline README.md:5]: Ranger-provided*

File: config.yaml [hunk 7]
```diff
@@ -14,18 +14,33 @@ options:
     description: |
       The secret Trino uses to communicate between nodes.
     type: string
-  k8s-tls-cert-name:
-    description: |
-      Name of TLS certificate as K8s secret
-    default: "not-provided"
-    type: string
+  https-enabled:
```
Comment (kelkawi-a) [inline config.yaml:17]: Is this config value still needed or is it being automatically set? If not, we can also remove it from ```CONTRIBUTING.md``` and ```README.md```.

File: README.md [hunk 8]
```diff
@@ -1,11 +1,45 @@
...
+Note: currently only Google Oauth authentication is supported.
+For information on how to set this up on Google see [here](https://developers.google.com/identity/protocols/oauth2).
+
+To disable TLS:
+```
+# disable https:
+juju config trino-k8s https-enabled=false
+
+# remove relation:
+juju remove-relation trino-k8s tls-certificates-operator --force
```
Comment (kelkawi-a) [inline README.md:39]: The ```--force``` flag is usually needed to ignore an error when removing the relation. Is there an error here? If not, we can remove the flag.

**Review by kelkawi-a — APPROVED** [review]
LGTM with a few comments. Great work!

File: README.md [hunk 9]
```diff
@@ -1,11 +1,45 @@
...
-Trino is a distributed SQL query engine designed to query large data sets distributed over one or more heterogeneous data sources.
+The operator charm comes with features such as:
+- Access control management supported with Ranger provided ACLs
 
+## Relations
+### tls-certificates
+The `tls-certificates` interface is used with the `tls-certificates-operator` charm. 
 
-## Other resources
+Note: The TLS settings here are for self-signed-certificates which are not recommended for production clusters, the tls-certificates-operator charm offers a variety of configurations, read more on the TLS charm [here](https://charmhub.io/tls-certificates-operator)
```
Comment (mertalpt) [inline README.md:11]: It might be more consistent to end this note with a period.

File: config.yaml (hunk 1 above)
Comment (AmberCharitos) [reply config.yaml]: Agreed, updated

File: src/charm.py (hunk 2 above)
Comment (AmberCharitos) [reply src/charm.py]: Updated

File: src/charm.py (hunk 3 above)
Comment (AmberCharitos) [reply src/charm.py]: Updated

File: src/charm.py (hunk 4 above)
Comment (AmberCharitos) [reply src/charm.py:316]: No, they were never required, just leftover from the original skeleton

File: src/tls.py
Comment (AmberCharitos) [reply src/tls.py:1]: Added, please take another look

File: README.md (hunk 5 above)
Comment (AmberCharitos) [reply README.md]: Thanks, updated

File: README.md (hunk 8 above)
Comment (AmberCharitos) [reply README.md]: No errors, removed the flag thanks

File: src/charm.py [hunk 10]
```diff
@@ -130,11 +138,11 @@ def _validate_db_conn_params(
                 f"connect-database: invalid database type {db_type!r}"
             )
 
-        if container.exists(f"/etc/trino/catalog/{db_name}.properties"):
+        if container.exists(f"{CATALOG_PATH}/{db_name}.properties"):
             raise ValueError(f"connect-database: {db_name!r} already exists!")
```
Comment (gtato) [inline src/charm.py:142]: what's "!r" ?

**Review by gtato — APPROVED** [review]

**Review by mertalpt — APPROVED** [review]
LGTM overall; some nitpicks and one somewhat important comment.

File: src/charm.py [hunk 11]
```diff
@@ -244,20 +252,81 @@ def _on_restart(self, event):
...
+            RuntimeError: In case keystore does not exist
+        """
+        google_id = self.config.get('google-client-id')
+        google_secret = self.config.get('google-client-secret')
+
+        if google_id is None or google_secret is None:
+            raise ValueError("Oauth credentials not provided")
+
+        path = f"{CONF_PATH}/keystore.p12"
+        if (container.exists(path) is False):
```
[the remark this answers is not in the document]
Comment (AmberCharitos) [reply src/charm.py]: Good point, thank you for this. I have updated usage of this throughout the charm

File: README.md (hunk 9 above)
Comment (AmberCharitos) [reply README.md]: Updated

File: requirements.txt [hunk 12]
```diff
@@ -1,2 +1,4 @@
 ops >= 2.2.0
-Jinja2==3.1.1
\ No newline at end of file
+Jinja2==3.1.1
```
[the remark this answers is not in the document]
Comment (AmberCharitos) [reply requirements.txt:2]: It has been changed, there was a trailing space after the original file. This was removed in this PR

**Pull request merged**